pkg/monitor/agent_test.go

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitor
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/fake"
+	cgtesting "k8s.io/client-go/testing"
+
+	"antrea.io/antrea/pkg/agent/config"
+	interfacestoretest "antrea.io/antrea/pkg/agent/interfacestore/testing"
+	openflowtest "antrea.io/antrea/pkg/agent/openflow/testing"
+	"antrea.io/antrea/pkg/agent/querier"
+	"antrea.io/antrea/pkg/apis/crd/v1beta1"
+	fakeclientset "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+	ovsconfigtest "antrea.io/antrea/pkg/ovs/ovsconfig/testing"
+	queriertest "antrea.io/antrea/pkg/querier/testing"
+)
+
+const ovsVersion = "2.10.0"
+
+func TestSyncAgentCRD(t *testing.T) {
+	existingCRD := &v1beta1.AntreaAgentInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "testAgentCRD",
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum: 1,
+		},
+		APIPort: 0,
+	}
+	partiallyUpdatedCRD := &v1beta1.AntreaAgentInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crdName,
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum: 0,
+		},
+		APIPort: 0,
+	}
+	entirelyUpdatedCRD := &v1beta1.AntreaAgentInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crdName,
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum: 0,
+		},
+		APIPort: 10349,
+	}
+	t.Run("partial update-success", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		monitor := newAgentMonitor(clientset, t)
+		monitor.agentCRD = existingCRD
+		monitor.querier.GetAgentInfo(partiallyUpdatedCRD, true)
+		monitor.syncAgentCRD()
+		crd, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Get(context.TODO(), "testAgentCRD", metav1.GetOptions{})
+		assert.NoError(t, err)
+		for i := range partiallyUpdatedCRD.AgentConditions {
+			partiallyUpdatedCRD.AgentConditions[i].LastHeartbeatTime.Time = crd.AgentConditions[i].LastHeartbeatTime.Time
+		}
+		assert.Equal(t, partiallyUpdatedCRD, crd)
+	})
+	t.Run("partial update-failure", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		clientset.PrependReactor("update", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &v1beta1.AntreaAgentInfo{}, errors.New("error updating agent crd")
+		})
+		monitor := newAgentMonitor(clientset, t)
+		monitor.agentCRD = existingCRD
+		monitor.syncAgentCRD()
+		_, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Update(context.TODO(), existingCRD, metav1.UpdateOptions{})
+		assert.Equal(t, "error updating agent crd", err.Error())
```

Comment:
I am not 100% sure this is useful (same for tests below). This is not really testing functionality, this is just testing your reactor above. You can keep it, but I would suggest adding a clarifying comment to avoid confusion.

Reply:
fixed